### PR TITLE
Added typestring parser checker

### LINKED ISSUE #1394: Addition of typestring parser checker
## Description of proposed feature

This particular package can actually check what type of typestring you should be using in your functions while writing the documentations. This could be really useful for the beginners like me and others or even contributors who write docs a lot as they can directly check it in the IDE.

## How can the new feature be used?

So first we have to download it via pip using `pip install typestring-parser` then we could use it as illustrated in the example given below.

For example:
```python
>>> from typestring_parser import parse
>>>
>>> parse('int')
<class 'int'>
>>> parse('int or str')
typing.Union[int, str]
>>> parse('list of str or str')
typing.Union[typing.List[str], str]
>>> parse('list of (int, str)')
typing.List[typing.Tuple[int, str]]
```

I was thinking if we could could have included in the manim package and include the usage in our docs

## Additional comments

See https://github.com/Dominik1123/typestring-parser for more info about this.

### PR DESCRIPTION
**Pull request by Nikhil0504**
## Changelog / Overview

Added `typestring-parser`

## Motivation

Now the contributors can check if their typestrings are correct more easily.
## Explanation for Changes

This could lessen the reviews and can PR's can be done faster. For first-timers, this would save them a lot of time as they could check here directly without going through many websites on whether they are correct or not.
## Further Comments

Closes #1394 
## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [ ] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by Nikhil0504** [comment]
@MrMallIronmaker can you check what the error is?

**Comment by MrMallIronmaker** [comment]
Nikhil - Thanks for the work you've been doing with the repo the past couple of days. I wanted to mention a couple things about communication, though.

First, I don't think I needed a ping on this issue. While I did reply to the original Discord comment, I don't have any special knowledge about doctypes or the module. Any member of the dev team could be equally suited for this question.

Second, I expected a more informative question than "can you check what the error was." Depending on how much experience you have with Github, CI, RTD, etc, I think good questions would be "How can I see what the problem was?" or "Why is RTD trying to run the code in my docs?" or "It works on my machine, why is it different here?" All of these questions are much more specific, so they show that you've put in the legwork in trying to teach yourself, they inform me a lot better about your experience so far, and they help me compose a better answer to your question.

I do appreciate your enthusiasm, and this message is one of those social norms to learn as you're getting started with FOSS and community software development, and in particular development with Manim.

File: docs/source/contributing/documentation.rst
Comment (kolibril13) [inline docs/source/contributing/documentation.rst:322]:
```suggestion
    As a helper for tool for typesets, you can use
```

**Comment by Nikhil0504** [comment]
I am actually getting an error because `typestring-parser` isn't being recognized in the docs as it is not being downloaded via `pip`. 
Maybe someone can tell where should I put this package to be installed as it is not being install from `requirements.txt`

Logs:
```
File "contributing/documentation.rst", line ?, in default
Failed example:
    from typestring_parser import parse
Exception raised:
    Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/doctest.py", line 1337, in __run
        compileflags, 1), test.globs)
      File "<doctest default[0]>", line 1, in <module>
        from typestring_parser import parse
    ModuleNotFoundError: No module named 'typestring_parser
```

**Comment by MrMallIronmaker** [comment]
Oh sorry, I realized I misspoke in my message earlier. I still think we should not add a dependency for a single section of the docs. The solution would not be live code to be run by the docstring engine but rather a handful of examples in the text itself.

For future reference, the right place to add it to the list is the pyproject.toml file.

File: docs/requirements.txt
Comment (MrMallIronmaker) [inline docs/requirements.txt:5]: It looks like this line didn't actually change; can you checkout the old requirements.txt so that this file doesn't have a change here?

File: docs/source/contributing/documentation.rst
Comment (MrMallIronmaker) [inline docs/source/contributing/documentation.rst:94]: What's the reason this changed? I'm not familiar with .rst type documentation.

**Review by MrMallIronmaker — CHANGES_REQUESTED** [review]
This addition looks good, let me know when you address the smaller changes.

File: docs/source/contributing/documentation.rst
Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst:94]: Apparently that is how we should write the note block in .rst files
https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html

Comment (kolibril13) [reply docs/source/contributing/documentation.rst:94]: Yep, `.. note::` is right, see also  here: https://manimce--1389.org.readthedocs.build/en/1389/contributing/documentation.html#note

Comment (MrMallIronmaker) [reply docs/source/contributing/documentation.rst:94]: Great, thanks for cleaning up this document.

**Comment by MrMallIronmaker** [comment]
It looks like "testsetup" gets rendered into the document. Is this intentional?

Screenshot from 2021-04-27 12-58-13

**Comment by RickyC0626** [comment]
> It looks like "testsetup" gets rendered into the document. Is this intentional?

I don't think so, maybe removing `Example` will help.

File: docs/source/contributing/documentation.rst
Comment (behackl) [inline docs/source/contributing/documentation.rst:327]: You just want a code block, right? Then this should do. Indentation of the block content seems to be missing though.

```suggestion
   Example::
```

Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst]: Ok, thank you.

**Review by MrMallIronmaker — APPROVED** [review]
lgtm, i can squash and merge this in a bit

File: docs/source/contributing/documentation.rst
Comment (naveen521kk) [inline docs/source/contributing/documentation.rst:337]: this doesn't render properly with syntax highlighting
image

Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst:338]: No let's not have it as it just makes the CI fail.

Comment (MrMallIronmaker) [reply docs/source/contributing/documentation.rst:337]: I would think there's some way to have a code block highlighted in the documentation, but not run - right? @naveen521kk 

If not, this is definitely not a breaking change.

Comment (naveen521kk) [reply docs/source/contributing/documentation.rst:337]: @Nikhil0504 I know how to do this *without breaking* things, but I want you to search or rather explore to find things out. 
You can easily add syntax highlighting, browse through though how it is done in other parts of manim itself or look at the reStructuredText Primer from sphinx or search the internet.

**Review by MrMallIronmaker — CHANGES_REQUESTED** [review]
requesting changes to enable syntax highlighting

File: docs/source/contributing/documentation.rst
Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst:336]: Hey, I tried fixing the issue and couldn't find a solution for it. Is it possible for you to fix it?

Comment (Darylgolden) [reply docs/source/contributing/documentation.rst:336]: Take a look at how https://docs.manim.community/en/latest/tutorials/quickstart.html does it

**Comment by Nikhil0504** [comment]
Fixed the code block issue at last.

File: docs/source/contributing/documentation.rst
Comment (naveen521kk) [inline docs/source/contributing/documentation.rst:326]: how is the below code `bash`? It's python.

Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst:326]: We use it under bash or cmd so it's not python right?

Comment (naveen521kk) [reply docs/source/contributing/documentation.rst:326]: Again, see how this is done in other places. It's obviously not `bash`.

Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst]: I made another change and checked locally which passed but over here the pre commit is failing

**Comment by Nikhil0504** [comment]
All checks have passed and I removed the `bash` Code which was changed to a `doctest ` block

File: docs/source/contributing/documentation.rst
Comment (naveen521kk) [inline docs/source/contributing/documentation.rst:328]: this line does not render correctly
image

Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst:328]: If I leave space for it, the CI failed (see https://github.com/ManimCommunity/manim/pull/1402/commits/d6036a0046ff5bdcc9769822f9a8826352495e18)
So I don't know what to do..

Comment (naveen521kk) [reply docs/source/contributing/documentation.rst:328]: You should disable doc test for that block of code.

Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst:328]: But then it will fail as this package is not in the dependency right?

Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst]: image

Comment (naveen521kk) [reply docs/source/contributing/documentation.rst]: That's what I wanted you to do :).

File: docs/source/contributing/documentation.rst
Comment (naveen521kk) [inline docs/source/contributing/documentation.rst:322]: can you like to `typestring-parser`?

File: docs/source/contributing/documentation.rst
Comment (naveen521kk) [inline docs/source/contributing/documentation.rst:323]: downloading seems to be wrong here.
```suggestion
   be accessed by first installing it via ``pip`` - ``pip install typestring-parser`` and
```

File: docs/source/contributing/documentation.rst
Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst:322]: the pip website link for it or the github page?

File: docs/source/contributing/documentation.rst
Comment (Nikhil0504) [reply docs/source/contributing/documentation.rst:323]: Ya I will change it

File: docs/source/contributing/documentation.rst
Comment (naveen521kk) [inline docs/source/contributing/documentation.rst:322]: rather than pasting the link near `typestring-parser` make that clickable.

**Review by naveen521kk — CHANGES_REQUESTED** [review]

**Review by MrMallIronmaker — APPROVED** [review]
docs are looking good to me.

**Review by naveen521kk — APPROVED** [review]
sgtm

**Comment by MrMallIronmaker** [comment]
@Nikhil0504 we can't merge until it's up-to-date, and we can't update the branch. Did you disable edits from maintainers on this branch?

**Comment by Nikhil0504** [comment]
I gave the access and made the branch up to date.

**Pull request merged**
